debezium-core/src/main/java/io/debezium/connector/common/BaseSourceTask.java

```diff
@@ -157,14 +164,49 @@ public final void start(Map<String, String> props) {
         }
 
         try {
-            final List<SourceRecord> records = doPoll();
-            logStatistics(records);
-            return records;
+            return pollRecords();
         }
         catch (RetriableException e) {
             stop(true);
             throw e;
         }
+        catch (ConnectException connectException) {
+            // In case of Kafka's ConnectException, retry starting the connector with a backoff strategy
+            Configuration config = Configuration.from(props);
+            int maxRetries = config.getInteger(CommonConnectorConfig.CONNECT_ERROR_MAX_RETRIES, 0);
+
+            if (maxRetries == 0) {
+                throw connectException;
+            }
+            else {
+                int initialRetryDelay = config.getInteger(CommonConnectorConfig.CONNECT_ERROR_INITIAL_RETRY_DELAY, 300);
+                int maxRetryDelay = config.getInteger(CommonConnectorConfig.CONNECT_ERROR_MAX_RETRY_DELAY, 10000);
+                DelayStrategy delayStrategy = DelayStrategy.exponential(initialRetryDelay, maxRetryDelay);
+
+                int totalRetries = 0;
+                boolean startedSuccessfully = false;
+                while (!startedSuccessfully) {
+                    try {
+                        totalRetries++;
+                        LOGGER.trace("Starting connector, attempt {}", totalRetries);
+                        stop(true);
+                        start(props);
+                        startedSuccessfully = true;
+                    }
+                    catch (Exception e) {
+                        LOGGER.error("Can't start the connector:", e);
+                        if (totalRetries == maxRetries) {
+                            LOGGER.info("Max retries to connect exceeded, stopping connector...");
+                            throw e;
+                        }
+                    }
+
+                    delayStrategy.sleepWhen(!startedSuccessfully);
+                }
```

Comment:
So I'm concerned this collides with the retrying logic in Kafka Connect itself (see [KIP 298](https://cwiki.apache.org/confluence/display/KAFKA/KIP-298%3A+Error+Handling+in+Connect)). I remember now also why we don't have a maximum retry count or duration, all that is handled by Kafka Connect. I.e. with KC, all we need to do is wrapping the retriable exception types in a `RetriableException`, and KC will keep trying as per its configuration for max retry duration and backoff period.

So instead of within `BaseSourceTask`, for Debezium Server we should provide an equivalent retrying implementation in the embedded engine code for Debezium Server. Ideally, the configuration options would be the same as for KC (i.e. `errors.retry.timeout` and `errors.retry.delay.max.ms`). WDYT?

Reply:
Worth checking it, I'll delve into it and let you know.

Reply:
Thanks, @gunnarmorling, that makes perfectly sense.
So basically what you are saying is moving the retry mechanism from the `BaseSourceTask` to `EmbeddedEngine`, right?
I can change the properties as suggested, but I think we still need `maxRetries` in Debezium Server in case it's not deployed within k8s or by any other HA mechanism. Since this is not handled by KC in this scope (i.e, Debezium Server), the RetriableException 

Something like this: (`EmbeddedEngine` from line 806`)
`                            catch (RetriableException e) {
                                LOGGER.info("Retrieable exception thrown, connector will be restarted", e);
                                // Retriable exception should be ignored by the engine
                                // and no change records delivered.
                                // The retry is handled in io.debezium.connector.common.BaseSourceTask.poll()
                            }
                            catch (ConnectException e) {
                                int initialRetryDelay = config.getInteger(CommonConnectorConfig.CONNECT_ERROR_INITIAL_RETRY_DELAY, 300);
                                int maxRetryDelay = config.getInteger(CommonConnectorConfig.CONNECT_ERROR_MAX_RETRY_DELAY, 10000);

                                DelayStrategy delayStrategy = DelayStrategy.exponential(initialRetryDelay, maxRetryDelay);

                                int maxRetries = config.getInteger(CommonConnectorConfig.CONNECT_ERROR_MAX_RETRIES, 0);

                                int totalRetries = 0;
                                boolean startedSuccessfully = false;
                                while (!startedSuccessfully) {
                                    try {
                                        totalRetries++;
                                        LOGGER.trace("Starting connector, attempt {}", totalRetries);
                                        task.stop();
                                        task.start(taskConfigs.get(0));
                                        startedSuccessfully = true;
                                    }
                                    catch (Exception ex) {
                                        LOGGER.error("Can't start the connector:", ex);
                                        if (totalRetries == maxRetries) {
                                            LOGGER.info("Max retries to connect exceeded, stopping connector...");
                                            throw ex;
                                        }
                                    }

                                    delayStrategy.sleepWhen(!startedSuccessfully);
                                }
                            }`

This seems working well and won't affect the connectors (and their tests).
WDYT? is that what you intedned?

Reply:
We can use it as in KC (for infinite retries support) - this way we will also have two properties instead of 3:
[errors.retry.timeout](https://docs.confluent.io/platform/current/installation/configuration/connect/sink-connect-configs.html#sinkconnectorconfigs_errors.retry.timeout)
The maximum duration in milliseconds that a failed operation will be reattempted. The default is 0, which means no retries will be attempted. Use -1 for infinite retries.

Reply:
I think that the `RetriableException` caught in the `EmbeddedEngine` won't start the Connector, so I'd apply the retry mechanism for both `ConnectException` and `RetriableException` in `EmbeddedEngine`.
Seems to me that the `BaseSourceTask` stops the connector and throws the `RetriableException` that is handled by KC automatically (based on the configuration of their retry) but in Debezium Server it's just skipped and as result the process is terminated, see below.

Current code:
```
 catch (RetriableException e) {
      LOGGER.info("Retrieable exception thrown, connector will be restarted", e);
      // Retriable exception should be ignored by the engine
      // and no change records delivered.
      // The retry is handled in io.debezium.connector.common.BaseSourceTask.poll()
 }
```

@gunnarmorling  Does it make sense? please share

Reply:
> So basically what you are saying is moving the retry mechanism from the BaseSourceTask to EmbeddedEngine, right?

> Seems to me that the BaseSourceTask stops the connector and throws the RetriableException that is handled by KC automatically (based on the configuration of their retry) but in Debezium Server it's just skipped and as result the process is terminated, see below.

Yes, exactly.